File: Project5/haskell/task5.hs
{- Program that automatically detects garbage collections from the timing information
 in Haskell (Haskell has automatic memory management)

Svita Kiran
CS333
11/13/2025
-}

import Control.DeepSeq (deepseq, NFData)
import System.CPUTime
import Text.Printf

-- allocate memory
allocateMemory :: Int -> [Int]
allocateMemory n = [1..n]

-- time function call
timeAction :: NFData a => IO a -> IO (a, Double)
timeAction action = do
    start <- getCPUTime
    result <- action
    result `deepseq` return ()  -- force evaluation
    end <- getCPUTime
    let diff = fromIntegral (end - start) / 1e12 :: Double  -- seconds
    return (result, diff)

-- iterations
iterations :: Int
iterations = 100

main :: IO ()
main = do
    let size = 1000000
    times <- mapM (\_ -> snd <$> timeAction (return $ allocateMemory size)) [1..iterations]
    let avgTime = sum times / fromIntegral iterations
    putStrLn $ "average time: " ++ show avgTime ++ " s"

    mapM_ (\(i :: Int, t :: Double) ->
               if t > 2 * avgTime
               then printf "iteration %d: %.6f s\n" i t
               else return ()) (zip [1..] times)

File: Project5/go/task5.go
/*
 * Program that automatically detects garbage collections from the timing information
 in Go (Go has automatic memory management)
 * 
 * Svita Kiran
 * 11/13/2025
 * CS333
*/

package main

import (
    "fmt"
    "time"
)

func allocateMemory() {
    _ = make([]int, 10_000_000)
}

func main() {
    const runs = 100
    times := make([]time.Duration, runs)

    for i := 0; i < runs; i++ {
        start := time.Now()
        allocateMemory()
        elapsed := time.Since(start)
        times[i] = elapsed
    }

    var total time.Duration
    for _, t := range times {
        total += t
    }
    avg := total / runs

    fmt.Printf("average time: %v\n", avg)

    for i, t := range times {
        if t > 2*avg {
            fmt.Printf("long run #%d: %v\n", i+1, t)
        }
    }
}
